netket/graph/lattice.py

```diff
@@ -104,323 +107,349 @@ def get_edges(atoms_positions, cutoff, distance_atol=cutoff_tol):
-            [
-                _np.matmul(_np.linalg.inv(self._basis_vectors.T), atom_coord)
-                for atom_coord in atoms_coord
-            ]
-        )
-        if (
-            atoms_coord_fractional.min() < -cutoff_tol
-            or atoms_coord_fractional.max() > 1 + cutoff_tol
-        ):
-            # Maybe there is another way to state this. I want to avoid that there exists the possibility that two atoms from different cells are at the same position:
-            raise ValueError(
-                "atoms must reside inside their corresponding unit cell, which includes only the 0-faces in fractional coordinates."
-            )
-        uniques = _np.unique(atoms_coord, axis=0)
-        if len(atoms_coord) != uniques.shape[0]:
-            atoms_coord = _np.asarray(uniques)
-            warnings.warn(
-                f"Some atom positions are not unique. Duplicates were dropped, and now atom positions are {atoms_coord}",
-                UserWarning,
-            )
+        self._primitives = self._clean_primitives(primitives)
+        self._ndim = self._primitives.shape[1]
 
-        self._atoms_coord = atoms_coord
+        self._basis, basis_fractional = self._clean_basis(basis, self._primitives)
+        self._pbc = self._clean_pbc(pbc, self._ndim)
 
-        if isinstance(pbc, bool):
-            self._pbc = [pbc] * self._basis_vectors.shape[1]
-        elif (
-            not isinstance(pbc, list)
-            or len(pbc) != self._basis_vectors.shape[1]
-            or sum([1 for pbci in pbc if isinstance(pbci, bool)])
-            != self._basis_vectors.shape[1]
-        ):
-            raise ValueError(
-                "pbc must be either a boolean or a list of booleans with the same dimension as the vector space you chose."
-            )
-        else:
-            self._pbc = pbc
+        self.extent = _np.asarray(extent, dtype=int)
 
-        extent = _np.asarray(extent)
-        self.extent = extent
-
-        atoms, cellANDlabel_to_site = create_points(
-            self._basis_vectors, extent, atoms_coord_fractional, pbc
+        sites, self._cell_coord_to_site = create_sites(
+            self._primitives, self.extent, basis_fractional, pbc
         )
         edges = get_true_edges(
-            self._basis_vectors, atoms, cellANDlabel_to_site, extent, distance_atol
+            self._primitives,
+            sites,
+            self._cell_coord_to_site,
+            self.extent,
+            distance_atol,
         )
         graph = _nx.MultiGraph(edges)
 
-        # Rename atoms
-        old_nodes = sorted(set([node for edge in edges for node in edge]))
-        self._atoms = [atoms[old_node] for old_node in old_nodes]
-        self._coord_to_site = {
-            tuple(atom["r_coord"]): new_site
-            for new_site, atom in enumerate(self._atoms)
-        }
+        # Rename sites
+        old_nodes = sorted(set(node for edge in edges for node in edge))
```

Comment:
Could you explain me what this does or why this indexing is better than indexing sites in some consistent manner (which would allow them to be used with convolutional networks for instance)?

Reply:
I left this logic as it was before.

My favorite ordering would be lexicographic in terms of the _cell coordinates_, so that, e.g.,
```python
>>> g.cell_coords
array([[0, 0, 0],
       [0, 0, 1],
       [0, 0, 2],
       [0, 1, 0],
       [0, 1, 1],
       [0, 1, 2],
       [0, 2, 0],
...
       [3, 3, 1],
       [3, 3, 2]])
```
The current code seems to do that usually, though I am not sure whether that is guaranteed. I'd be happy to standardize the cell coord ordering.

Reply:
What do the columns mean in this matrix?

> The current code seems to do that usually, 

Then I'm probably happy with it too, I don't mind what the exact order is, as long as it has some disoverable and documentable logic.

> though I am not sure whether that is guaranteed. 

It's filtered through two libraries we have no control over, I wouldn't take anything guaranteed about them...

> I'd be happy to standardize the cell coord ordering and make that more explicit in the code.

That would be nice :) I think it really only requires specifying `id`s when creating the sites

Reply:
From what I can see, this is indeed the ordering provided by the current code; I've added a test to make sure it doesn't accidentally break.